### ISSUE TITLE
SDCA .Fit() never returns when running in an ASP.NET MVC Full .NET Framework 4.6.1 / 4.7.2

### ISSUE THREAD

**Issue by Ianpwest** [dotnet/machinelearning#3830]
When running the sentiment analysis sample this line never returns. No exception is thrown.
`// STEP 4: Train the model fitting to the DataSet
            ITransformer trainedModel = trainingPipeline.Fit(trainingData);`

This is not the case in .net core web apps or .NET Framework 4.6.1 console apps. Only MVC.

You can recreate the issue by creating a brand new ASP.NET MVC .NET Framework 4.6.1 Web Application. Changing to target x64. Installing the Microsoft.ML nuget and running the sentiment analysis code.

Thanks.

**Comment by RahulKumar4** [comment]
Hi,
I am also facing the same issue. I am creating a nuget package which runs on net472. 

running the sentiment analysis sample this line never returns. No exception is thrown.
 ITransformer trainedModel = trainingPipeline.Fit(trainingData);

**Comment by antoniovs1029** [comment]
Hi, @Ianpwest @RahulKumar4 . I have tried to reproduce your issue (both in ASP.NET and by creating a nuget package, both on .Net Framework 4.7.2, and the ASP.NET also on .Net Framework 4.6.1), but I've actually gotten an exception right on the .Fit() method you're pointing to, while running this on debug mode. So I am not sure if I'm doing something wrong when replicating the issue, and I would appreciate if you could share a .zip including a project that would reproduce this.

I got the exception below using .NET Framework as I've mentioned, while running in debug mode. But I didn't get it on ASP.NET Core. So this is consistent with what the original poster said, and I will assume this exception is actually what caused the issue in here. It is about not being able to find a DLL that should be there after installing ML.NET nuget. Since I am not very familiar with ASP.NET apps, I will still have to confirm if this exception wasn't thrown because I made some mistake, and that's why I'd appreciate your repo.

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Shuffle input cursor reader failed with an exception
  Source=Microsoft.ML.Data
  StackTrace:
   at Microsoft.ML.Transforms.RowShufflingTransformer.Cursor.MoveNextCore()
   at Microsoft.ML.Data.RootCursorBase.MoveNext()
   at Microsoft.ML.Trainers.TrainingCursorBase.MoveNext()
   at Microsoft.ML.Trainers.SdcaTrainerBase`3.TrainCore(IChannel ch, RoleMappedData data, LinearModelParameters predictor, Int32 weightSetCount)
   at Microsoft.ML.Trainers.StochasticTrainerBase`2.TrainModelCore(TrainContext context)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.TrainTransformer(IDataView trainSet, IDataView validationSet, IPredictor initPredictor)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.Fit(IDataView input)
   at Microsoft.ML.Data.EstimatorChain`1.Fit(IDataView input)
   at WebApplication472.Controllers.SentimentRun.Run() in C:\Users\anvelazq\source\repos\WebApplication472\WebApplication472\Controllers\Class1.cs:line 49
   at WebApplication472.Controllers.HomeController.About() in C:\Users\anvelazq\source\repos\WebApplication472\WebApplication472\Controllers\HomeController.cs:line 19
   at System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters)
   at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c.<BeginInvokeSynchronousActionMethod>b__9_0(IAsyncResult asyncResult, ActionInvocation innerInvokeState)
   at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult)
   at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.End()
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass11_0.<InvokeActionMethodFilterAsynchronouslyRecursive>b__0()
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass11_2.<InvokeActionMethodFilterAsynchronouslyRecursive>b__2()

Inner Exception 1:
InvalidOperationException: Splitter/consolidator worker encountered exception while consuming source data

Inner Exception 2:
DllNotFoundException: Unable to load DLL 'CpuMathNative': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
```

**Comment by antoniovs1029** [comment]
So I've been able to remove the exception on my last post by manually copying the CpuMathNative.dll file into the bin folder where the app is running. Now I am able to reproduce your issue, as the .Fit() method isn't returning even after some minutes.

It seems the exception I was getting is related to the fact that my newly created ASP.NET Framework app is using the Package.config format instead the PackageReference format, and that cause the DLL to not be copied. I am still unsure on how/if I can use PackageReference on ASP.NET Framework, so for the time being I'll simply continue to look into this by manually copying that file. So if you, @Ianpwest or @RahulKumar4 have a repro project that doesn't need to do this hack, I'll greatly appreciate it. Thanks!

**Comment by antoniovs1029** [comment]
Hi, @Ianpwest  and @RahulKumar4 . Updating to ML.NET version `1.5.0-preview1` or -`preview2` fixes this issue.

I was able to reproduce this issue with ML.NET version 1.0 (which is probably the version you were using at the time this issue got created) and 1.4.0 (which is the last non-preview version that ML.NET has released). But it is solved on preview of 1.5.

Also, for the record, I noticed that this issue only happened when using a SDCA trainer (such as the one on the sample); this worked correctly if I changed the trainer to any other Binary classification trainer. This is because this issue was actually caused by this lines of code on the SDCA trainer, waiting for the algorithm to converge, and got stuck while waiting for that ([code here](https://github.com/dotnet/machinelearning/blob/214926fcf5753ad62acf32c0759bdaf8fcd13b73/src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs#L594-L654)).
